### DISCUSSION TITLE
What do async/await keywords really do?

### DISCUSSION THREAD

**Discussion by ShervanN** [ProfessionalCSharp/ProfessionalCSharp2021#174] [Q&A] [answered]
Hi,

The snippet code is from Page 460, Please see the two images. What do async/await keywords really do?
With await keyword, the type of data variable is SomeData, without await keyword is ValueTask<SomeData>

ReadAsync signature is:
```
  public virtual ValueTask<T> ReadAsync(CancellationToken cancellationToken = default) 
{
//....
}

```
![vs01](https://user-images.githubusercontent.com/108445707/225753205-631621c1-4df5-4850-94f8-af9f0570048d.png)
![vs02](https://user-images.githubusercontent.com/108445707/225753219-431706d6-cce9-4321-8639-f2deeb5641f6.png)

**Comment by christiannagel** [comment, marked as answer]
## ValueTask

The `await` keyword works with every returned **awaiter** object that implements the `GetAwaiter` method. When async/await was introduced, the `await` keyword required a `Task`.

Like the `Task` class, the `ValueTask` implements the `GetAwaiter` method, and thus `await` can wait on `ValueTask` as well. The difference is that `Task` is a class and thus an object is created that needs cleanup from the garbage collector, whereas `ValueTask` is a struct. Compared to an asynchronous invocation, an object is a small overhead. However, if you get a lengthy list, e.g. with asynchronous streaming, not every returned item is really asynchronous, thus the `ValueTask` can reduce the overhead.

The `Task` class returns a `TaskAwaiter`, the `ValueTask` returns a `ValueTaskAwaiter`.

## async and return values

With the sample code, `reader.ReadAsync` returns a `ValueTask<SomeData>`. As you can see from your first image, without using `async`, this is what's returned. You can use this `ValueTask` as you want and do some parallel processing before you check for the result.

Using the `await` keyword, `await` creates a code block with all the code lines that follow in the method - including the assignment to the variable `data`, and passes this code block to the `OnCompleted(Action)` method to the awaiter. When the awaiter completes, this code block is invoked, and first the result of the task is assigned to the `data` variable.

That's why you see a different result type with the `await`. If a method returns `Task<string>` or `ValueTask<string>`, you'll see a string variable assigned when using `await`. A library could return `MyCustomTask<SomeData>`. In case `MyCustomTask` implements the `GetAwaiter` method, you can do `SomeData data = await MethodReturningCustomTask()`.

I hope this helps.

More:

Stephen Toub has a great blog post: [How Async/Await Really Works in C#](https://devblogs.microsoft.com/dotnet/how-async-await-really-works/)

**Reply by ShervanN** [reply]
Thank you so much.
